gh-101000: Add os.path.splitroot()

barneygale opened the pull request:
This PR introduces `os.path.splitroot()`. See #101000 for motivation.

In `ntpath`, the implementation derives from `splitdrive()`. The `splitdrive()` function now calls `splitroot()`, and returns `drive, root + tail`. Other functions now call `splitroot()` rather than `splitdrive()`. In most cases this replaces their own parsing of path roots. It also avoids adding a stack frame.

In `posixpath`, the `normpath()` function now calls `splitroot()` rather than parsing path roots itself.

In `pathlib`, path constructors now call `splitroot()` rather than using a slow OS-agnostic implementation. Performance:

```shell
$ ./python -m timeit -s 'from pathlib import PureWindowsPath' 'PureWindowsPath("C:/", "foo", "bar")'
50000 loops, best of 5: 6.04 usec per loop  # before
50000 loops, best of 5: 4.03 usec per loop  # after
$ ./python -m timeit -s 'from pathlib import PurePosixPath' 'PurePosixPath("/", "etc", "hosts")'
100000 loops, best of 5: 3.11 usec per loop  # before
100000 loops, best of 5: 2.77 usec per loop  # after
```

Future work:

- Improve performance by using native `nt._path_splitroot()`

<!-- gh-issue-number: gh-101000 -->
* Issue: gh-101000
<!-- /gh-issue-number -->

barneygale committed:
gh-101000: Add os.path.splitroot()

barneygale committed:
Use splitroot() from pathlib

barneygale committed:
Use splitroot() from posixpath

barneygale committed:
Use splitroot() from ntpath

barneygale committed:
Optimizations

h-vetinari commented on Lib/posixpath.py:
```diff
@@ -135,6 +135,35 @@ def splitdrive(p):
+    contains anything after the root. For example:
+
+        splitdrive('foo/bar') == ('', '', 'foo/bar')
+        splitdrive('/foo/bar') == ('', '/', 'foo/bar')
```
```suggestion
        splitroot('foo/bar') == ('', '', 'foo/bar')
        splitroot('/foo/bar') == ('', '/', 'foo/bar')
```

I'd also add an example with the two-slashes case

h-vetinari commented on Lib/posixpath.py:
```diff
@@ -135,6 +135,35 @@ def splitdrive(p):
     return p[:0], p
```
IIUC the intention not to let `splitdrive` wrap `splitroot` for `posixpath` is because this is so simple and therefore faster?
Perhaps:
```suggestion
    # intentionally not wrapping splitroot for performance reasons;
    # use empty slice to match str/bytes type for drive to that of p
    return p[:0], p
```

barneygale replied:
As the docstring mentions, the drive is always empty on POSIX, so the implementation of `splitdrive()` is trivial. I'm not sure if this really needs calling out in a comment personally, nor the fact that we preserve the data type (*all* functions in `os.path` do this!)

barneygale replied:
I'm resolving this thread because I feel the implementation and docstring are sufficiently explanatory. Hope that makes sense!

h-vetinari commented on Lib/ntpath.py:
```diff
@@ -765,8 +785,8 @@ def relpath(path, start=None):
     try:
         start_abs = abspath(normpath(start))
         path_abs = abspath(normpath(path))
-        start_drive, start_rest = splitdrive(start_abs)
-        path_drive, path_rest = splitdrive(path_abs)
+        start_drive, _, start_rest = splitroot(start_abs)
+        path_drive, _, path_rest = splitroot(path_abs)
```
This looks like a behaviour change to me (depending on whether the now-ignored root matches between `start_abs` and `path_abs`) -- I'm guessing it's intentional? Or are those corner-cases somehow already leading to the same outcome (or known-wrong currently)?

eryksun replied:
With `splitdrive()`, the root would be at the beginning of `start_rest`. However, the root doesn't matter because the values are further processed into a list of components as follows:

```python
        start_list = [x for x in start_rest.split(sep) if x]
        path_list = [x for x in path_rest.split(sep) if x]
```

h-vetinari replied:
ah, I had seen that list comprehension but misread the fact that it will also strip the separator. 👍

h-vetinari commented on Lib/ntpath.py:
```diff
@@ -525,14 +550,9 @@ def normpath(path):
+        drive, root, path = splitroot(path)
 
-        comps = path.split(sep)
+        comps = path.lstrip(sep).split(sep)
```
I'm surprised that this needs an additional `lstrip(sep)`, since that's exactly what `splitroot` should now be putting into `root`?

eryksun replied:
The path isn't normalized yet. For example, say it's `"D:\\\\spam\\eggs"`, which splits as `('D:', '\\', '\\spam\\eggs')`.

h-vetinari replied:
Ah right, before that was in the now-removed if-branch. 👍

h-vetinari commented on Lib/ntpath.py:
```diff
@@ -212,15 +239,13 @@ def split(p):
     Either part may be empty."""
     p = os.fspath(p)
     seps = _get_bothseps(p)
-    d, p = splitdrive(p)
+    d, r, p = splitroot(p)
     # set i to index beyond p's last slash
     i = len(p)
     while i and p[i-1] not in seps:
         i -= 1
     head, tail = p[:i], p[i:]  # now tail has no slashes
-    # remove trailing slashes from head, unless it's all slashes
-    head = head.rstrip(seps) or head
-    return d + head, tail
+    return d + r + head.rstrip(seps), tail
```
This makes me wonder if `splitroot` shouldn't ensure that `tail` does not start with slashes anymore...?

I guess I'm not getting something, because this behaviour is tested explicitly, but I'm wondering what that's for...

eryksun replied:
Given `p` is path-like, it's guaranteed that `type(fspath(p))().join(splitroot(p)) == fspath(p)`.

The `head` and `tail` values here come from splitting the value of `p` that's assigned by `d, r, p = splitroot(p)`. Say the input path is "D:/eggs//spam". Then `head` is set to "eggs//", and `tail` is "spam". The return value is `('D:/eggs', 'spam')`.

barneygale replied:
If I were writing this function fresh, I'd probably name the `tail` variable `name` or `base` or something like that, so that it doesn't conflict with the documented return value of `splitdrive()` / `splitroot()`. But I'd rather not change the naming of these variables now unless a core dev asks me to, as it adds headaches for `git blame` users.

h-vetinari commented on Lib/ntpath.py:
```diff
@@ -169,35 +171,60 @@ def splitdrive(p):
+
+        splitroot('//server/share/') == ('//server/share', '/', '')
+        splitroot('C:/Users/Barney') == ('C:', '/', 'Users/Barney')
+        splitroot('Windows') == ('', '', 'Windows')
```
Perhaps add an example that redundant slashes in tail do not get removed.

h-vetinari reviewed (commented):
Drive-by comments from seeing this PR on discourse, feel free to disregard if I'm saying/asking something stupid.

barneygale committed:
Correct and expand examples in splitroot() docstring

eryksun commented on Lib/ntpath.py:
```diff
@@ -169,35 +171,61 @@ def splitdrive(p):
+
+        splitroot('//server/share/') == ('//server/share', '/', '')
+        splitroot('C:/Users/Barney') == ('C:', '/', 'Users/Barney')
+        splitroot('C:///spam///ham') == ('C:', '/', '//spam///egg')
```
```suggestion
        splitroot('C:///spam///ham') == ('C:', '/', '//spam///ham')
```

barneygale replied:
![image](https://user-images.githubusercontent.com/960340/212396564-cf985c7a-b66f-44ea-ae76-45ea45130243.png)

barneygale committed:
Update Lib/ntpath.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

barneygale committed:
Use splitroot() from pathlib.PurePath.with_name()

barneygale committed:
Reduce ntpath.normpath() diff noise

AlexWaygood commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,26 @@ the :mod:`glob` module.)
+
+   Split the pathname *path* into a triad ``(drive, root, tail)`` where:
+
+   1. *drive* is an optional mount point, exactly like :func:`splitdrive`;
```
I sympathise with the desire to keep repetition out of the docs, but I think it'll be friendlier to readers if we restate the description of _drive_ here, rather than forcing them to click through to the `splitdrive` function docs to find out when _drive_ will be an empty string or not.

Also, as a "typing guy" at heart -- "optional" has a special meaning for users of typing, where it means "could be `None`. I like the more explicit "Could be something meaningful, or the empty string" description in the `splitdrive()` docs.

barneygale committed:
Simplify ntpath.commonpath() now that 'isabs' is unused.

AlexWaygood commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,26 @@ the :mod:`glob` module.)
+   2. *root* is an optional sequence of separators following the drive; and
+   3. *tail* is anything after the root.
+
+   On Posix, *drive* is always empty. The *root* may be empty (relative path),
```
```suggestion
   On Posix systems, *drive* is always empty. The *root* may be empty (if it is a relative path),
```

AlexWaygood commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,26 @@ the :mod:`glob` module.)
+   3. *tail* is anything after the root.
+
+   On Posix, *drive* is always empty. The *root* may be empty (relative path),
+   a single forward slash (absolute path), or two forward slashes
```
```suggestion
   a single forward slash (if it is an absolute path), or two forward slashes
```

AlexWaygood commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,26 @@ the :mod:`glob` module.)
+
+   On Posix, *drive* is always empty. The *root* may be empty (relative path),
+   a single forward slash (absolute path), or two forward slashes
+   (implementation-defined per the POSIX standard).
```
What's the "POSIX standard"? Can that be linked to?

eryksun replied:
POSIX stands for Portable Operating System Interface ([POSIX.1 FAQ](https://www.opengroup.org/austin/papers/posix_faq.html)). The following is the relevant statement from the [pathname resolution](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13) discussion in IEEE Std 1003.1-2017:

> If a pathname begins with two successive \<slash\> characters, the first component following the leading \<slash\> characters may be interpreted in an implementation-defined manner, although more than two leading \<slash\> characters shall be treated as a single \<slash\> character.

AlexWaygood commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,26 @@ the :mod:`glob` module.)
+   a single forward slash (absolute path), or two forward slashes
+   (implementation-defined per the POSIX standard).
+
+   On Windows, *drive* may be a UNC sharepoint or a traditional DOS drive. The
```
This is the first use of the term "DOS" in this document -- can we link to a definition of the term?

AlexWaygood commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,26 @@ the :mod:`glob` module.)
+
+   In all cases, ``drive + root + tail`` will be the same as *path*.
+
+   .. versionadded:: 3.12
```
It would be nice to add a short doctest example, similar to `os.path.splitdrive` immediately above. Also: I think this deserves an entry in "What's New in Python 3.12"!

AlexWaygood reviewed (commented):
I'm persuaded that this is a good idea. Thanks for working on this!

Here's a docs review. Haven't got to looking at the implementation yet (will do soon).

barneygale committed:
Reduce posixpath.normpath() diff noise

barneygale committed:
Improve documentation

barneygale committed:
Add whatsnew entry.

barneygale committed:
Simplify ntpath.splitroot() slightly

eryksun commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,38 @@ the :mod:`glob` module.)
+
+   Split the pathname *path* into a triad ``(drive, root, tail)`` where:
+
+   1. *drive* is a mount point or the empty string;
```
```suggestion
   1. *drive* is a device name, mount point or the empty string;
```

The `spltdrive()` documentation says that a drive is a mount point, but that's not generally true. A mount point is a path that's mounted by a filesystem directory. A UNC share is a mount point for a redirected or virtual filesystem, but a drive such as "D:" is usually the name of a volume device. The automatic mount point is the root path "D:\\", assuming the volume is even formatted.

Mapped drives and substitute drives are logical device names, though they can behave more like symlinks. For an example of the latter, if substitute drive "S:" targets "\\??\\C:\\Windows\\System32", then a relative symlink "S:\\link.exe" -> "..\\notepad.exe" resolves to "C:\\Windows\\notepad.exe". If "S:" were a real device name, then trying to traverse "S:\\link.exe" would fail as invalid.

As a pure path operation, we don't check the device type or whether it exists. Any device name is accepted as a drive, including known serial devices such as "\\\\.\\NUL", "\\\\.\\CON" and "\\\\.\\COM1".

eryksun commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,38 @@ the :mod:`glob` module.)
+      ('', '/', 'etc/hosts')
+
+   On Windows, *drive* may be a UNC sharepoint or a traditional drive-letter drive. The
+   *root* may be empty, a forward slash, or a backward slash. For example::
```
```suggestion
   On Windows, *drive* may be a drive-letter name, a UNC share, or a device
   name. The *root* may be empty, a forward slash, or a backward slash. For
   example::
```

eryksun commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,38 @@ the :mod:`glob` module.)
+      >>> splitroot('//server/share/')
+      ('//server/share', '/', '')
+      >>> splitroot('C:/Users/Barney')
+      ('C:', '/', 'Users/Barney')
```
```suggestion
      >>> splitroot('C:/Users/Barney')
      ('C:', '/', 'Users/Barney')
      >>> splitroot('//server/share/')
      ('//server/share', '/', '')
      >>> splitroot('//?/Volume{12345678-1234-1234-1234-123456781234}/')
      ('//?/Volume{12345678-1234-1234-1234-123456781234}', '/', '')
```

barneygale committed:
Apply suggestions from code review

Co-authored-by: Eryk Sun <eryksun@gmail.com>

eryksun commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,41 @@ the :mod:`glob` module.)
+
+   On POSIX systems, *drive* is always empty. The *root* may be empty (if *path* is
+   relative), a single forward slash (if *path* is absolute), or two forward slashes
+   (implementation-defined per `IEEE Std 1003.1 2013 Edition; 4.13 Pathname Resolution
```
```suggestion
   (implementation-defined per `IEEE Std 1003.1-2017; 4.13 Pathname Resolution
```

barneygale committed:
Update Doc/library/os.path.rst

Co-authored-by: Eryk Sun <eryksun@gmail.com>

barneygale committed:
Note that drive may be empty on Windows

barneygale committed:
Re-order drive example

eryksun commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,41 @@ the :mod:`glob` module.)
+   For example::
+
+      >>> splitroot('/etc/hosts')
+      ('', '/', 'etc/hosts')
```
```suggestion
      >>> splitroot('home/barney')
      ('', '', 'home/barney')
      >>> splitroot('/home/barney')
      ('', '/', 'home/barney')
      >>> splitroot('//home/barney')
      ('', '//', 'home/barney')      
      >>> splitroot('///home/barney')
      ('', '/', '//home/barney')
```

barneygale replied:
Perhaps getting too long IMHO. @AlexWaygood what do you think? (and below)

eryksun replied:
The example with a relative path could be skipped as something most people will understand intuitively. I think it's helpful to show "/" vs "//" vs "///", or at least "//" vs "///". I think the result with "//" may be surprising because Linux and macOS handle it the same as "/" and "///".

AlexWaygood replied:
No other function in this document has a code block with more than two examples, so I agree that four examples here would be too many.

The last example _is_ somewhat surprising (to me at least), so I'd definitely keep that one. Perhaps the second could be skipped.

eryksun replied:
If I had to choose just two Unix examples, I'd opt for the following:

```
      >>> splitroot('/home/barney')
      ('', '/', 'home/barney')
      >>> splitroot('//home/barney')
      ('', '//', 'home/barney')
```

With three examples, I'd add the triple slash example: 

```
      >>> splitroot('///home/barney')
      ('', '/', '//home/barney')
```

I prefer for the examples to be similar in terms of component names, which emphasizes at a glance how the paths are syntactically different in terms of slashes.

AlexWaygood replied:
I'll defer to yours and Barney's superior expertise on paths as to which specific ones are kept, but four is definitely too many for me :)

I'd prefer two, but I think I could tolerate three, if necessary.

barneygale replied:
I've opted for 3 so we can include the triple-slash case, which I agree is surprising to newcomers.

eryksun commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,41 @@ the :mod:`glob` module.)
+   example::
+
+      >>> splitroot('Windows/notepad')
+      ('', '', 'Windows/notepad')
```
```suggestion
      >>> splitroot('Users/Barney')
      ('', '', 'Users/Barney')
      >>> splitroot('/Users/Barney')
      ('', '/', 'Users/Barney')
```

AlexWaygood replied:
This doesn't seem necessary to me -- I think it can be inferred from the example immediately below. We should keep the docs here relatively terse, in keeping with the rest of the `os.path` documentation.

eryksun replied:
If I had to choose just two Windows examples, I'd choose the following:

```
      >>> splitroot('C:/Users/Barney')
      ('C:', '/', 'Users/Barney')
      >>> splitroot('//Server/Share/Users/Barney')
      ('//Server/Share', '/', 'Users/Barney')
```

With 3 examples, I'd add a "//?/UNC" device example:

```
      >>> splitroot('//?/UNC/Server/Share/Users/Barney')
      ('//?/UNC/Server/Share', '/', 'Users/Barney')
```

barneygale replied:
I've opted for 2 examples in this case as I think the `//?/UNC` behaviour should be obvious enough if you're using those sorts of paths.

eryksun commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,41 @@ the :mod:`glob` module.)
+      >>> splitroot('//server/share/')
+      ('//server/share', '/', '')
+      >>> splitroot('//?/Volume{12345678-1234-1234-1234-123456781234}/')
+      ('//?/Volume{12345678-1234-1234-1234-123456781234}', '/', '')
```
The GUID device example is too long. Switch to UNC.

```suggestion
      >>> splitroot('//?/UNC/server/share/')
      ('//?/UNC/server/share', '/', '')
```

barneygale committed:
Update Doc/library/os.path.rst

Co-authored-by: Eryk Sun <eryksun@gmail.com>

barneygale committed:
Adjust docstring examples

AlexWaygood commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,41 @@ the :mod:`glob` module.)
+   Split the pathname *path* into a triad ``(drive, root, tail)`` where:
+
+   1. *drive* is a device name, mount point or the empty string;
+   2. *root* is a sequence of separators following the drive or the empty string; and
```
I don't like the word "sequence" here — my immediate reaction is "what kind of sequence? A list? A tuple? A deque?" That said, I'm struggling to think of something better :(

barneygale replied:
perhaps "a string of separators"?

eryksun replied:
```suggestion
   2. *root* is either one or more separators that follow the drive or the
      empty string; and
```

AlexWaygood replied:
Or perhaps:

```suggestion
   2. *root* is either a string of one or more separators that follow the drive, or
      the empty string; and
```

barneygale replied:
Personally I don't think "one or more" is necessary; we get into the details in the following paragraphs. The "more" could also give the impression that _any number_ of separators are included in the root, which isn't the case.

```suggestion
   2. *root* is a string of separators that follow the drive, or the empty string; and
```

AlexWaygood commented on Lib/ntpath.py:
```diff
@@ -816,21 +837,19 @@ def commonpath(paths):
-            isabs, = set(p[:1] == sep for d, p in drivesplits)
-        except ValueError:
-            raise ValueError("Can't mix absolute and relative paths") from None
+        if len(set(r for d, r, p in drivesplits)) != 1:
```
Since we're touching this line anyway, we may as well modernise the code slightly by using a set comprehension

```suggestion
        if len({r for d, r, p in drivesplits}) != 1:
```

AlexWaygood commented on Lib/ntpath.py:
```diff
@@ -816,21 +837,19 @@ def commonpath(paths):
         # now otherwise type errors for mixing strings and bytes would not be
         # caught.
-        if len(set(d for d, p in drivesplits)) != 1:
+        if len(set(d for d, r, p in drivesplits)) != 1:
```
```suggestion
        if len({d for d, r, p in drivesplits}) != 1:
```

AlexWaygood reviewed (commented):
Still haven't had a chance to review the code properly (will do soon!), but here's a few drive-by comments in the meantime

barneygale committed:
Apply suggestions from code review

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

eryksun commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,41 @@ the :mod:`glob` module.)
 
+.. function:: splitroot(path)
+
+   Split the pathname *path* into a triad ``(drive, root, tail)`` where:
```
This would the only use of ["triad"](https://en.wiktionary.org/wiki/triad#Noun) in the source code and documentation. I'm used to either ["3-tuple"](https://en.wiktionary.org/wiki/-tuple#Suffix) or ["triple"](https://en.wiktionary.org/wiki/triple#Noun) in this context.

```suggestion
   Split the pathname *path* into a 3-tuple of strings ``(drive, root, tail)`` where:
```

AlexWaygood replied:
Oh, I quite like "triad" here — for me it's quite clear and succinct.

barneygale replied:
I picked "triad" rather than "triple" because `splitdrive()` uses "pair" rather than "double". But honestly I don't mind "triple" if that's more obvious.

I don't like `3-tuple` because the bit in monospace already makes it obvious that it's a tuple. We could make that work by adjusting it as follows:

> Split the *path* into a 3-item tuple of strings *drive*, *root* and *tail*, where:

I'm borrowing "n-item tuple" from the `urllib.parse` docs as I have them to hand.

barneygale replied:
```suggestion
   Split the pathname *path* into a triple ``(drive, root, tail)`` where:
```

AlexWaygood replied:
I like "3-item tuple" or "triad". Not so keen on "3-tuple" or "triple".

eryksun replied:
When I search the web for "triad" used with "tuple", I do come across "triad" to mean a 3-item ordered sequence. So maybe it's just me. It's not as common as "triple" or "3-tuple", however. I also like "3-item tuple".

The term "tuple" is a generalization of the words double, triple, quadruple, quintuple, and so on, which are based on Latin words that combine a number (e.g. duo, tres, quattor, quintus) and "-plus" (i.e. English "-fold").

barneygale replied:
The docs for `os.walk()` use "3-tuple" and "triple". There's plenty of other occurences of "triple"; less of "3-tuple".

What don't you like about "triple", Alex?

AlexWaygood replied:
I've seen "triple" used in the CPython docs in a few other places, but I've never seen it used as a noun outside the CPython docs. I know it _can_ be used as a noun, but I think technical documentation should generally strive to be clear to laypeople, not just people who are used to reading technical documentation. (There are obviously limits to how far you should take that principle, but it seems like a good principle to aspire towards.) Part of that, in my opinion, means using words in the same way that people will be used to seeing them used in contexts outside of technical documentation.

AlexWaygood replied:
Anyway, it sounds like we all like "3-item tuple", so why don't we just go with that? :)

barneygale replied:
```suggestion
   Split the pathname *path* into a 3-item tuple ``(drive, root, tail)`` where:
```

eryksun commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,41 @@ the :mod:`glob` module.)
+      >>> splitroot('//home/eryk')
+      ('', '//', 'home/eryk')
+      >>> splitroot('///home/eryk')
+      ('', '/', '//home/eryk')
```
This is your idea and PR, and an extension of your ongoing work on pathlib. Please credit yourself (e.g. "/home/barney"). You definitely deserve it.

barneygale replied:
Still, it feels a bit self-serving to plaster my name all over the `os.path` docs! I'll pick a neutral name.

barneygale committed:
Update Doc/library/os.path.rst

barneygale committed:
Change example username in docs to 'Sam'

barneygale committed:
Adjust first paragraph to use prose

... and not belabour the fact that the empty string may be returned as
any/all items.

AlexWaygood commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,40 @@ the :mod:`glob` module.)
+   <https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13>`_.)
+   For example::
+
+      >>> splitroot('/home/sam')
+      ('', '/', 'home/sam')
+      >>> splitroot('//home/sam')
+      ('', '//', 'home/sam')
+      >>> splitroot('///home/sam')
+      ('', '/', '//home/sam')
+
+   On Windows, *drive* may be empty, a drive-letter name, a UNC share, or a device
+   name. The *root* may be empty, a forward slash, or a backward slash. For
+   example::
+
+      >>> splitroot('C:/Users/Sam')
+      ('C:', '/', 'Users/Sam')
+      >>> splitroot('//Server/Share/Users/Sam')
+      ('//Server/Share', '/', 'Users/Sam')
```
Maybe some of these examples should use `bytes` instead of strings, since the function works with `bytes` just as well.

Also, does the function work with `os.PathLike` objects? If so, should we state that explicitly? If not, should it?

eryksun replied:
Support for `bytes` and `os.PathLike` objects is mentioned in the first paragraph of the `os.path` documentation.

> This module implements some useful functions on pathnames. To read or write files see [open()](https://docs.python.org/3/library/functions.html#open), and for accessing the filesystem see the [os](https://docs.python.org/3/library/os.html#module-os) module. The path parameters can be passed as strings, or bytes, or any object implementing the [os.PathLike](https://docs.python.org/3/library/os.html#os.PathLike) protocol.

AlexWaygood replied:
None of the existing examples in this module use `bytes` examples, but I'd still be tempted to have one or two of the examples here use `bytes` paths. It can't hurt, in my opinion:

```suggestion
      >>> splitroot('/home/sam')
      ('', '/', 'home/sam')
      >>> splitroot('//home/sam')
      ('', '//', 'home/sam')
      >>> splitroot(b'///home/sam')
      (b'', b'/', b'//home/sam')

   On Windows, *drive* may be empty, a drive-letter name, a UNC share, or a device
   name. The *root* may be empty, a forward slash, or a backward slash. For
   example::

      >>> splitroot('C:/Users/Sam')
      ('C:', '/', 'Users/Sam')
      >>> splitroot(b'//Server/Share/Users/Sam')
      (b'//Server/Share', b'/', b'Users/Sam')
```

barneygale replied:
Byte paths are a niche tool on POSIX. They're used because POSIX pathnames aren't guaranteed to conform to any particular encoding, and so highly portable applications will use `bytes` to avoid guessing.

I'm not aware of any use case for them on Windows. AFAIK they're supported only for conformity with POSIX.

I think we'd be drawing undeserved attention to them if we include them in the examples. It also makes the differences between the examples harder to grok, I think.

AlexWaygood replied:
Alright, let's leave this as it is then! Thanks!

barneygale replied:
Hope that's alright! Thanks.

AlexWaygood commented on Doc/library/os.path.rst:
```diff
@@ -488,6 +488,40 @@ the :mod:`glob` module.)
+      >>> splitroot('//Server/Share/Users/Sam')
+      ('//Server/Share', '/', 'Users/Sam')
+
+
```
Should only have one newline here, rather than two

```suggestion
```

AlexWaygood commented on Lib/ntpath.py:
```diff
@@ -169,35 +171,61 @@ def splitdrive(p):
+    elif normp[1:2] == colon:
+        if normp[2:3] == sep:
+            # Absolute drive-letter path, e.g. X:\Windows
+            return p[:2], p[2:3], p[3:]
```
This could be written slightly more simply using the `partition` string method, but perhaps it's better to use slicing here for consistency with the other parts of this function. I'll leave it up to you:

```suggestion
            return p.partition(sep)
```

barneygale replied:
That won't work I'm afraid! `p` contains both forward and backward slashes, so partitioning on `sep` but not `altsep` will miss forward slashes.

barneygale replied:
Note how there are two variables in play here: `p` is the original path, and `normp` has all its forward slashes converted to backward slashes. We use `normp` to find the indices of path separators, and then slice `p` using those indices.

AlexWaygood replied:
Ah I see! In that case, could we add some tests that would fail if we used the incorrect `partition()` code? It looks like if I make this change locally, a `join()` test and a `splitdrive()` test start to fail, but none of the tests for `splitroot()` specifically start to fail.

AlexWaygood commented on Lib/posixpath.py:
```diff
@@ -135,6 +135,35 @@ def splitdrive(p):
+        return empty, empty, p
+    elif p[1:2] != sep or p[2:3] == sep:
+        # Absolute path, e.g.: '/foo', '///foo', '////foo', etc.
+        return empty, sep, p[1:]
```
Could similarly use `str.partition()` here as well:

```suggestion
        return p.partition(sep)
```

barneygale replied:
I think that could obscure what the code is doing slightly. Firstly because the code above and below doesn't use `partition()`, and secondly because the astute reader might ask "why use partition() if you already know the position of the separator?" and conclude they're missing something.

AlexWaygood replied:
Yes, this one isn't nearly as much of a "simplification" as the other. Better to leave it as you have it currently, as you say :))

AlexWaygood reviewed (changes requested):
I believe there are currently no tests that `os.path.splitroot` works with `os.PathLike` objects. Just trivial tests like this should do fine, but we should make sure it's tested:

https://github.com/python/cpython/blob/0e75a55b3f696e4481a51c7e414503aecf85caba/Lib/test/test_ntpath.py#L946-L947

barneygale committed:
Update Doc/library/os.path.rst

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

barneygale committed:
Add tests for bytes (POSIX only) and path-like objects (both platforms)

AlexWaygood commented on Misc/NEWS.d/next/Library/2023-01-12-21-22-20.gh-issue-101000.wz4Xgc.rst:
```diff
@@ -0,0 +1,2 @@
+Add :func:`os.path.splitroot()`, which splits a path into a triad of
+``(drive, root, tail)``.
```
I wonder if the NEWS entry should also mention the `pathlib` speedup; it seems significant.

barneygale committed:
Add tests for mixed path separators (Windows only)

barneygale committed:
Remove errant newline.

AlexWaygood reviewed (commented):
This is looking really good to me now, and I'm very close to hitting "approve". My only concern (other than my comment about the NEWS entry) is that I'm still not sure the test coverage is quite there. It looks like the tests for `splitdrive()` account for a lot of edge cases that aren't really tackled in the tests for `splitroot()` yet, e.g.

https://github.com/python/cpython/blob/e99e3cd2262918ee4148258f3f49360ee85850f2/Lib/test/test_ntpath.py#L117-L119

and 

https://github.com/python/cpython/blob/e99e3cd2262918ee4148258f3f49360ee85850f2/Lib/test/test_ntpath.py#L129-L130

It's true that, since `splitdrive()` now uses `splitroot()`, these edge cases are in some sense already covered -- the tests for `splitdrive()` will start failing if a bug is introduced to `splitroot()` at some later date in the future. But it will be highly confusing if the tests for `splitdrive()` start failing, yet the tests for `splitroot()` all still pass, when the bug is actually in the implementation for `splitroot()`.

barneygale committed:
Move most test cases from `test_splitdrive` to `test_splitroot`

barneygale committed:
Mention pathlib performance improvement in news entry.

AlexWaygood reviewed (approved):
Looks great to me. Thanks, as ever, for your patience and perseverance!

@eryksun, Any further comments from you? :)

AlexWaygood committed:
Merge branch 'main' into gh-101000-splitroot
